File: solutions/sonnet-1/sonnet/typescript/lcs.ts
function main(input: string) {
  // Split into lines, handling both \n and \r\n, but preserve up to 2 lines
  // even if one or both are empty.
  const lines = input.split(/\r?\n/);
  const a = lines[0] ?? "";
  const b = lines[1] ?? "";

  const n = a.length;
  const m = b.length;

  let prev = new Uint32Array(m + 1);
  let curr = new Uint32Array(m + 1);

  for (let i = 1; i <= n; i++) {
    const ca = a.charCodeAt(i - 1);
    for (let j = 1; j <= m; j++) {
      if (ca === b.charCodeAt(j - 1)) {
        curr[j] = prev[j - 1] + 1;
      } else {
        curr[j] = prev[j] > curr[j - 1] ? prev[j] : curr[j - 1];
      }
    }
    const tmp = prev;
    prev = curr;
    curr = tmp;
  }

  console.log(prev[m].toString());
}

let input = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => {
  input += chunk;
});
process.stdin.on("end", () => {
  main(input);
});

File: solutions/sonnet-1/sonnet/python/lcs.py
import sys


def main():
    data = sys.stdin.read().split("\n")
    a = data[0] if len(data) > 0 else ""
    b = data[1] if len(data) > 1 else ""

    n, m = len(a), len(b)
    if n < m:
        a, b = b, a
        n, m = m, n

    prev = [0] * (m + 1)
    curr = [0] * (m + 1)

    for i in range(1, n + 1):
        ai = a[i - 1]
        for j in range(1, m + 1):
            if ai == b[j - 1]:
                curr[j] = prev[j - 1] + 1
            else:
                curr[j] = prev[j] if prev[j] >= curr[j - 1] else curr[j - 1]
        prev, curr = curr, prev

    print(prev[m])


if __name__ == "__main__":
    main()
